src/core/IncomingRequestManager.cpp

```diff
@@ -233,6 +378,51 @@ void IncomingContactRequest::setNickname(const QString &nickname)
     emit nicknameChanged();
 }
 
+#ifdef PROTOCOL_NEW
+void IncomingContactRequest::setChannel(Protocol::ContactRequestChannel *channel)
+{
+    if (connection) {
+        qDebug() << "Replacing connection on an IncomingContactRequest. Old connection is" << connection->age() << "seconds old.";
+        connection->close();
+    }
+
+    // When the channel is closed, also close the connection
+    connect(channel, &Protocol::Channel::invalidated, this,
+        [this,channel]() {
+            if (connection == channel->connection() &&
+                connection->purpose() == Protocol::Connection::Purpose::InboundRequest)
+            {
+                qDebug() << "Closing connection attached to an IncomingContactRequest because ContactRequestChannel was closed";
+                connection->close();
+            }
+        }
+    );
+
+    qDebug() << "Assigning connection to IncomingContactRequest from" << m_hostname;
+    connection = channel->connection();
+    connection->setParent(this);
+    if (!connection->setPurpose(Protocol::Connection::Purpose::InboundRequest)) {
+        qDebug() << "Setting purpose on incoming contact request connection failed; killing connection";
```

Comment:
setPurpose itself calls BUG() for invalid transitions, so I think these should all be qWarning at most.

We really need a logging overhaul at some point too. Fortunately most of it is disabled for release builds.